src/device_manager.rs

```diff
@@ -0,0 +1,461 @@
+                        .is_some()
+                    {
+                        return idx;
+                    }
+                }
+                IoType::PhysicalMmio => continue,
+            }
+        }
+        resources.len()
+    }
+
+    // Unregister resources with all entries addresses valid.
+    fn unregister_resources(&mut self, resources: &[IoResource]) {
+        for res in resources.iter() {
+            // The resources addresses being unregistered is sucessfully allocated before.
+            let addr = res.try_unwrap();
+
+            match res.res_type {
+                IoType::Pio => self.pio_bus.remove(&Range(addr, res.size)),
+                IoType::Mmio => self.mmio_bus.remove(&Range(addr, res.size)),
+                IoType::PhysicalMmio => continue,
+            };
+        }
+    }
+
+    fn allocate_irq_resource(
+        &mut self,
+        interrupt: Option<IrqResource>,
+    ) -> Result<Option<IrqResource>> {
+        match interrupt {
+            Some(IrqResource(irq)) => {
+                // Allocate irq resource
+                let irq_num = self
+                    .resource
+                    .allocate_irq(irq)
+                    .map_err(Error::IrqAllocate)?;
+                Ok(Some(IrqResource(Some(irq_num))))
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn free_irq_resource(&mut self, interrupt: Option<IrqResource>) {
+        match interrupt {
+            Some(IrqResource(irq)) => self.resource.free_irq(irq),
+            None => return,
+        }
+    }
+
+    fn allocate_id_resource(&mut self) -> Result<u32> {
+        self.resource
+            .allocate_instance_id()
+            .map_err(Error::InstanceIdAllocate)
+    }
+
+    fn free_id_resource(&mut self, id: u32) {
+        self.resource.free_instance_id(id);
+    }
+
+    /// Register a new device with its parent bus and resources request set.
+    /// Return Ok(instance_id) when sucessfully registered for caller usage.
+    pub fn register_device(
+        &mut self,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: &mut Vec<IoResource>,
+        interrupt: Option<IrqResource>,
+    ) -> Result<(u32)> {
+        // Allocate an instance id
+        let id = self.allocate_id_resource()?;
+
+        // Reserve resources
+        if let Err(Error::IoResourceAllocate(idx, e)) = self.allocate_io_resources(resources) {
+            // Free allocated resources if one resource failed to allocate.
+            if idx > 0 {
+                self.free_io_resources(&resources[0..idx - 1]);
+                self.free_id_resource(id);
+                return Err(Error::IoResourceAllocate(idx, e));
+            }
+        }
+
+        // Register device resources
+        let register_len = self.register_resources(dev.clone(), resources);
+        // Unregister and free resources once failed.
+        if register_len < resources.len() && register_len > 0 {
```

Comment:
We should release io resources and id even if register_len is 0.

Reply:
Yep, will change it.